bmorph/core/models.py

```diff
@@ -0,0 +1,361 @@
+            ('normalize', MinMaxScaler())
+            ]),
+        'elevation': FunctionTransformer(elevation_scaler),
+        'temperature': StandardScaler(),
+        'contributing_area': MinMaxScaler(),
+        'other': StandardScaler(),
+        }
+    # Check if any custom transformers were given
+    for varname, transformer in custom_transformers.items():
+        transformers[varname] = transformer
+    # Fit the transformers on a per variable level
+    fit_transformers = {}
+    for key in df.columns:
+        # Look for key in transformers, otherwise use 'other' as default
+        fit_transformers[key] = transformers.get(key, 'other')
+        fit_transformers[key].fit(df[[key]])
+    return fit_transformers
+
+
+def apply_transformers(df: pd.DataFrame, fit_transformers: dict, inverse=False):
+    """
+    Applies transformers fit in the `fit_transformers` function.
+    This does not mutate data in place.
+
+    Parameters
+    ----------
+    df:
+        Dataframe with complete set of all data for training
+        This should have a MultiIndex with levels ('site', 'time')
+    fit_transformers:
+        Dictionary of transformers to be applied. They must be
+        generated by the ``fit_transformers`` function or manually
+        fit before this function can be used. Format is
+        ``{variable: fit_transformer}``
+    inverse:
+        Whether to invert from the given transformers. This simply
+        calls the ``inverse_transform`` method instead of ``transform``
+
+    Returns
+    -------
+    out:
+        Dataframe transformed by each of the transformers
+    """
+    out = pd.DataFrame(index=df.index)
+    if not inverse:
+        for key in df:
+            out[key] = fit_transformers[key].transform(df[[key]])
+    else:
+        for key in df:
+            out[key] = fit_transformers[key].inverse_transform(df[[key]])
+    return out
+
+
+def save_transformers(transformers: dict, path='.') -> List[str]:
+    """Saves transformers to disk"""
+    out_files = []
+    for name, tformer in transformers.items():
+        of = f'{path}{os.sep}{name}-bmorph-transformer.joblib'
+        dump(tformer, of)
+        out_files.append(of)
+    return out_files
+
+
+def load_transformers(file_list: List[str]) -> dict:
+    """Load transformers from disk"""
+    transformers = {}
+    for file in file_list:
+        name = file.split(os.sep)[-1].replace('.joblib', '').split('-')[0]
+        transformers[name] = load(file)
+    return transformers
+
+
+def split_train_test_sites(sites: List, train_frac: float=0.8) -> Tuple[List]:
+    """
+    Randomly partition sites into test and train samples.
+
+    Parameters
+    ----------
+    sites:
+        List of sites to partition
+    train_frac:
+        Fraction of sites to place into the training partition
+
+    Returns
+    -------
+    train_sites, test_sites
+    """
+    n_train = int(train_frac * len(sites))
+    shuffled = copy.copy(sites)
```

Comment:
Why not directly shuffle `sites`?

Reply:
Calling `random.shuffle` mutates the object in place, which would change the `sites` list. This introduces some pretty unintuitive behavior. So, we make a copy and shuffle that so that we are not messing with a user's variable.